osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
```

Comment:
Right, good question! Serializing schedule execution is a design choice. It allows osquery users to schedule to their hearts content. That schedule will never stress the daemon to starve CPU resources on a node given a potentially intense set of intervals.

We want to keep the remaining architecture simple, and that means no locking within the registry abstraction around the result of plugin calls.

Detecting re-entrancy and buffering is the correct approach. But some modifications are needed with the current implementation, but we're nearly there!

I'd suggest keeping a global-scope (or `static` member for organization/style) that keeps a map of `tid_t` to booleans representing servicing-db-work. When you enter a DB related method you obtain a locker instance that checks the map (in a thread-safe way): ` if (kMapOfTIDs[tid])`. If it's true then you have reentered, so buffer. If it's false set the `tid -> true`, save a local `i_locked_ -> true` and in the dtor, check if `i_locked` and subsequently `tid -> false`.